Shard fork choice rule (#1773)
ethereum/eth2.0-specs: specs/phase1/shard-fork-choice.md

    while True:
        children = [
            root for root in shard_store.blocks.keys()
            if shard_store.blocks[root].shard_parent_root == head_shard_root
        ]
        if len(children) == 0:
            return head_shard_root
        # Sort by latest attesting balance with ties broken lexicographically
        head_shard_root = max(
            children, key=lambda root: (get_shard_latest_attesting_balance(store, shard_store, root), root)
        )
```

#### `get_shard_ancestor`

```python
def get_shard_ancestor(store: Store, shard_store: ShardStore, root: Root, slot: Slot) -> Root:
    block = shard_store.blocks[root]
    if block.slot > slot:
        return get_shard_ancestor(store, shard_store, block.shard_parent_root, slot)
    elif block.slot == slot:
        return root
    else:
        # root is older than queried slot, thus a skip slot. Return earliest root prior to slot
        return root
```

### Handlers

#### `on_shard_block`

```python
def on_shard_block(store: Store, shard_store: ShardStore, signed_shard_block: SignedShardBlock) -> None:
    shard_block = signed_shard_block.message

>     shard = shard_store.shard

Reviewer: What do you guys think about adding `shard` to a `ShardBlock` field?   

I think it's going to get annoying computing or tracking the origin `shard` for a `ShardBlock` : /

Reviewer 2: yeah, I agree. being able to quickly validate baseline info about messages coming off the wire is valuable

    # 1. Check shard parent exists
    assert shard_block.shard_parent_root in shard_store.block_states
    pre_shard_state = shard_store.block_states[shard_block.shard_parent_root]

    # 2. Check beacon parent exists
    assert shard_block.beacon_parent_root in store.block_states
    beacon_state = store.block_states[shard_block.beacon_parent_root]

    # 3. Check that block is later than the finalized epoch slot (optimization to reduce calls to get_ancestor)

>     finalized_slot = compute_start_slot_at_epoch(store.finalized_checkpoint.epoch)

Reviewer 2: I think this is a bit off

Instead, it is probably something like
* get the beacon state for the latest finalized
* get the shard state for the shard in the latest finalized
* ensure shard_block.slot is later than the shard state slot in the latest finalized beacon state (this can certainly be for a slot earlier than the latest finalized beacon slot

    assert shard_block.slot > finalized_slot

    # 4. Check block is a descendant of the finalized block at the checkpoint finalized slot
    assert (

>         shard_block.beacon_parent_root == store.finalized_checkpoint.root

Reviewer 2: In the case that the beacon parent root is as old as possible (the latest finalized root), does the second condition below (with `get_ancestor`) not cover this case? Why is this base case needed?

            if (
                shard_blocks[root].shard_parent_root == head_shard_root
                and shard_blocks[root].slot > head_shard_state.slot
            )
        ]
        if len(children) == 0:
            return head_shard_root
        # Sort by latest attesting balance with ties broken lexicographically
        head_shard_root = max(
            children, key=lambda root: (get_shard_latest_attesting_balance(store, shard_store, root), root)
        )
```

#### `get_shard_ancestor`

```python
def get_shard_ancestor(store: Store, shard_store: ShardStore, root: Root, slot: Slot) -> Root:
    block = shard_store.blocks[root]
    if block.slot > slot:
        return get_shard_ancestor(store, shard_store, block.shard_parent_root, slot)
    elif block.slot == slot:
        return root
    else:
        # root is older than queried slot, thus a skip slot. Return earliest root prior to slot
        return root
```

### Handlers

#### `on_shard_block`

```python
def on_shard_block(store: Store, shard_store: ShardStore, signed_shard_block: SignedShardBlock) -> None:
    shard_block = signed_shard_block.message
    shard = shard_store.shard

>     # 1. Check shard parent exists

Reviewer: Add `assert shard_block.shard == shard` ?

Reviewer 2: I thought about this and assumed that since the client would be the one piping the shard block into this function with the associated `shard_store` that this base mapping already occurred.

That said, I suppose there is not hurt in double checking here

Reviewer: Fair point. I'll still add the check on Prysm just to be extra defensive in good way

Reviewer 2: I'm happy for it to go in.

Another way we can design this is to have `shard_stores` be a map of shard -> shard_store. And `on_shard_block` actually selects the `shard_store` to use.

Curious what @hwwhww feels about the options here

Reviewer 2: I see that.

- [Introduction](#introduction)
- [Fork choice](#fork-choice)
  - [Helpers](#helpers)
    - [`ShardStore`](#shardstore)
    - [`get_forkchoice_shard_store`](#get_forkchoice_shard_store)
    - [`get_shard_latest_attesting_balance`](#get_shard_latest_attesting_balance)
    - [`get_shard_head`](#get_shard_head)
    - [`get_shard_ancestor`](#get_shard_ancestor)
    - [`get_pending_shard_blocks`](#get_pending_shard_blocks)
  - [Handlers](#handlers)
    - [`on_shard_block`](#on_shard_block)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Introduction

This document is the shard chain fork choice spec for part of Ethereum 2.0 Phase 1. It assumes the [beacon chain fork choice spec](./fork-choice.md).

## Fork choice

### Helpers

#### `ShardStore`

```python
@dataclass
class ShardStore:
    shard: Shard
    blocks: Dict[Root, ShardBlock] = field(default_factory=dict)
    block_states: Dict[Root, ShardState] = field(default_factory=dict)
```

#### `get_forkchoice_shard_store`

```python
def get_forkchoice_shard_store(anchor_state: BeaconState, shard: Shard) -> ShardStore:
    return ShardStore(
        shard=shard,

>         blocks={anchor_state.shard_states[shard].latest_block_root: ShardBlock(slot=anchor_state.slot)},

Reviewer 2: add `shard=shard` to the anchor block stub

        block_states={anchor_state.shard_states[shard].latest_block_root: anchor_state.copy().shard_states[shard]},
    )
```

#### `get_shard_latest_attesting_balance`

```python
def get_shard_latest_attesting_balance(store: Store, shard_store: ShardStore, root: Root) -> Gwei:
    state = store.checkpoint_states[store.justified_checkpoint]
    active_indices = get_active_validator_indices(state, get_current_epoch(state))
    return Gwei(sum(
        state.validators[i].effective_balance for i in active_indices
        if (
            i in store.latest_messages
            # TODO: check the latest message logic: currently, validator's previous vote of another shard
            # would be ignored once their newer vote is accepted. Check if it makes sense.
            and store.latest_messages[i].shard == shard_store.shard
            and get_shard_ancestor(

>                 store, shard_store, store.latest_messages[i].root, shard_store.blocks[root].slot

Reviewer: ```suggestion
                store, shard_store, store.latest_messages[i].shard_root, shard_store.blocks[root].slot
```

I think that's `shard_root`.  Maybe 👇 is  better to avoid future confusions:

```python
class LatestMessage(object):
    epoch: Epoch
    beacon_root: Root
    shard: Shard
    shard_root: Root
```

Author: Nice catch! 🙈  Thanks!!

            ) == root
        )
    ))
```

#### `get_shard_head`

```python
def get_shard_head(store: Store, shard_store: ShardStore) -> Root:
    # Execute the LMD-GHOST fork choice

>     shard_blocks = shard_store.blocks

Reviewer 2: A pre-filter of the blocks might be clearer/cleaner than doing the `and shard_blocks[root].slot > head_shard_state.slot` check in each iteration of the `while` loop below

```suggestion
    shard_blocks = [shard_block for shard_block in shard_store.blocks if shard_block.slot > head_shard_state.slot
```

    head_beacon_root = get_head(store)
    head_shard_state = store.block_states[head_beacon_root].shard_states[shard_store.shard]
    shard_head_root = head_shard_state.latest_block_root
    while True:
        # Find the valid child block roots
        children = [
            root for root in shard_store.blocks.keys()
            if (
                shard_blocks[root].shard_parent_root == shard_head_root
                and shard_blocks[root].slot > head_shard_state.slot
            )
        ]
        if len(children) == 0:
            return shard_head_root
        # Sort by latest attesting balance with ties broken lexicographically
        shard_head_root = max(
            children, key=lambda root: (get_shard_latest_attesting_balance(store, shard_store, root), root)
        )
```

#### `get_shard_ancestor`

```python
def get_shard_ancestor(store: Store, shard_store: ShardStore, root: Root, slot: Slot) -> Root:
    block = shard_store.blocks[root]
    if block.slot > slot:
        return get_shard_ancestor(store, shard_store, block.shard_parent_root, slot)
    elif block.slot == slot:
        return root
    else:

>         # root is older than queried slot, thus a skip slot. Return earliest root prior to slot

Reviewer 2: We recently modified this comment a bit in the beacon fork choice

```suggestion
        # root is older than queried slot, thus a skip slot. Return most recent root prior to slot
```

        return root
```

#### `get_pending_shard_blocks`

```python
def get_pending_shard_blocks(store: Store, shard_store: ShardStore) -> Sequence[ShardBlock]:
    """

>     Return the shard blocks branch that from shard head to beacon head.

Reviewer 2: I think this might be a bit more clear but still not 100% happy with it

```suggestion
    Return the canonical shard block branch that has not yet been crosslinked.
```

    # Check shard parent exists
    assert shard_block.shard_parent_root in shard_store.block_states
    pre_shard_state = shard_store.block_states[shard_block.shard_parent_root]

    # Check beacon parent exists
    assert shard_block.beacon_parent_root in store.block_states

    # Check that block is later than the finalized shard state slot (optimization to reduce calls to get_ancestor)
    finalized_beacon_state = store.block_states[store.finalized_checkpoint.root]
    finalized_shard_state = finalized_beacon_state.shard_states[shard]
    assert shard_block.slot > finalized_shard_state.slot

    # Check block is a descendant of the finalized block at the checkpoint finalized slot
    finalized_slot = compute_start_slot_at_epoch(store.finalized_checkpoint.epoch)
    assert (
        get_ancestor(store, shard_block.beacon_parent_root, finalized_slot) == store.finalized_checkpoint.root
    )

    # Add new block to the store
    shard_store.blocks[hash_tree_root(shard_block)] = shard_block

    # Check the block is valid and compute the post-state
    beacon_head_root = get_head(store)
    beacon_head_state = store.block_states[beacon_head_root]

>     assert verify_shard_block_message(beacon_head_state, pre_shard_state, shard_block, shard_block.slot, shard)

Reviewer 2: Why do we pass in `shard_block.slot` as an independent variable? Is there ever a case that we want  the `slot` passed in to differ from `shard.slot`?

Author: Yes it seems leftover of outdated assumption. 👍 
I realized we can also remove `shard` since we got `block.shard`, so it's now just `verify_shard_block_message(beacon_parent_state, shard_parent_state, block)` 🎉

**Notice**: This document is a work-in-progress for researchers and implementers.

## Table of contents

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [Introduction](#introduction)
- [Fork choice](#fork-choice)
  - [Helpers](#helpers)
    - [Extended `Store`](#extended-store)
    - [Updated `get_forkchoice_store`](#updated-get_forkchoice_store)
    - [`get_shard_latest_attesting_balance`](#get_shard_latest_attesting_balance)
    - [`get_shard_head`](#get_shard_head)
    - [`get_shard_ancestor`](#get_shard_ancestor)
    - [`filter_shard_block_tree`](#filter_shard_block_tree)
    - [`get_filtered_block_tree`](#get_filtered_block_tree)
  - [Handlers](#handlers)
    - [`on_shard_block`](#on_shard_block)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Introduction

This document is the shard chain fork choice spec for part of Ethereum 2.0 Phase 1.

## Fork choice

### Helpers

#### Extended `Store`

```python
@dataclass

> class Store:

Author: Added independent `ShardStore`. It indeed makes the code cleaner. 👍

    return Gwei(sum(
        state.validators[i].effective_balance for i in active_indices
        if (

>             i in store.latest_messages and

Reviewer 2: In the rest of the repo, we've used the operation on the start of the next line (e.g. `and`) rather than the end of the current line.

Pretty sure this is a recent recommendation in PEP but can't currently find the standard

>             store.latest_messages[i].shard == shard_store.shard and

Reviewer 3: I don't understand this. It seems that only if the validator has attested to the shard _in their latest message_ that their vote would be counted. So in the next slot, when the attest to a different shard, their vote would be removed. Is that how it should work?

Reviewer 2: It would be in the next epoch rather than slot.
The beacon chain will generally continue to advance to provide a new basis for the shard, and if not, committees will continue to turn over and vote on the shard.

Even if we store multiple shard votes for a single validator, this condition is still important. We need to prevent using a vote on shard A to try to influence shard B (example: using your assigned vote from shard A to vote on a root from shard B, thus throwing your weight behind a shard you weren't assigned)

Reviewer 3: True but even if it's in the last epoch, it means the "old" blocks suddenly lose their attestations -- so the commitee in the next epoch could just ignore them if they haven't been crosslinked. Seems weird to me.

Reviewer 2: How about we flag this for discussion on Tuesday? I think it's worth hashing out in convo
